### ISSUE TITLE
Negative window geometry coordinates result in incorrect size

### ISSUE THREAD

**Issue by diniamo** [hyprwm/Hyprland#8495]
### Already reported ? *

- [X] I have searched the existing open and closed issues.

### Regression?

No

### System Info and Version

<details> <summary>System/Version info</summary>

```sh
Hyprland 0.45.0 built from branch  at commit 098e491a43e8b26f4382b48651a4131464bf6a2f  ().
Date: 2024-11-15
Tag: , commits: 
built against aquamarine 0.4.5


flags set:
debug


System Information:
System name: Linux
Node name: diniamo-PC
Release: 6.11.7
Version: #1-NixOS SMP PREEMPT_DYNAMIC Fri Nov  8 15:31:04 UTC 2024


GPU information: 
01:00.0 VGA compatible controller [0300]: NVIDIA Corporation TU116 [GeForce GTX 1660 Ti] [10de:2182] (rev a1) (prog-if 00 [VGA controller])
NVRM version: NVIDIA UNIX x86_64 Kernel Module  565.57.01  Thu Oct 10 12:29:05 UTC 2024


os-release: ANSI_COLOR="1;34"
BUG_REPORT_URL="https://github.com/NixOS/nixpkgs/issues"
BUILD_ID="24.11.20241115.c5a76ad"
CPE_NAME="cpe:/o:nixos:nixos:24.11"
DEFAULT_HOSTNAME="diniamo-PC"
DOCUMENTATION_URL="https://nixos.org/learn.html"
HOME_URL="https://nixos.org/"
ID=nixos
ID_LIKE=""
IMAGE_ID=""
IMAGE_VERSION=""
LOGO="nix-snowflake"
NAME=NixOS
PRETTY_NAME="NixOS 24.11 (Vicuna)"
SUPPORT_URL="https://nixos.org/community.html"
VARIANT=""
VARIANT_ID=""
VENDOR_NAME=NixOS
VENDOR_URL="https://nixos.org/"
VERSION="24.11 (Vicuna)"
VERSION_CODENAME=vicuna
VERSION_ID="24.11"


plugins:

======Config-Start======


Config File: /home/diniamo/.config/hypr/hyprland.conf: Read Succeeded
exec-once = /nix/store/ii86djqx6kb39qqhzsbwd5hmqf276xzz-dbus-1.14.10/bin/dbus-update-activation-environment --systemd --all && systemctl --user stop hyprland-session.target && systemctl --user start hyprland-session.target
animations {
  bezier=myBezier, 0.05, 0.9, 0.1, 1.1
  animation=windows, 1, 5, myBezier
  animation=windowsOut, 1, 5, default, popin 80%
  animation=windowsMove, 1, 5, default, popin 80%
  animation=fade, 1, 5, default
  animation=border, 1, 5, default
  animation=borderangle, 0, 8, default
  animation=workspaces, 0
  animation=specialWorkspace, 0
  enabled=true
}

binds {
  allow_workspace_cycles=true
  workspace_center_on=1
}

cursor {
  no_hardware_cursors=0
  use_cpu_buffer=true
}

decoration {
  blur {
    enabled=false
  }

  shadow {
    color=rgba(24273a99)
    enabled=false
  }
  rounding=0
}

dwindle {
  force_split=2
  pseudotile=true
  special_scale_factor=0.950000
}

general {
  border_size=1
  col.active_border=rgb(8aadf4)
  col.inactive_border=rgb(494d64)
  gaps_in=0
  gaps_out=0
  layout=dwindle
}

gestures {
  workspace_swipe=true
  workspace_swipe_cancel_ratio=0.150000
  workspace_swipe_direction_lock=false
}

group {
  groupbar {
    col.active=rgb(8aadf4)
    col.inactive=rgb(494d64)
    text_color=rgb(cad3f5)
  }
  col.border_active=rgb(8aadf4)
  col.border_inactive=rgb(494d64)
  col.border_locked_active=rgb(8bd5ca)
}

input {
  touchpad {
    natural_scroll=true
    scroll_factor=0.800000
    tap-to-click=false
  }
  accel_profile=flat
  kb_layout=hu
  kb_options=caps:swapescape,altwin:swap_lalt_lwin
  scroll_method=on_button_down
}

misc {
  animate_manual_resizes=true
  animate_mouse_windowdragging=true
  background_color=rgb(24273a)
  disable_autoreload=true
  disable_hyprland_logo=true
  disable_splash_rendering=true
  exit_window_retains_fullscreen=true
  focus_on_activate=true
  force_default_wallpaper=0
  key_press_enables_dpms=true
  middle_click_paste=false
  mouse_move_enables_dpms=false
  new_window_takes_over_fullscreen=1
}

opengl {
  nvidia_anti_flicker=false
}

render {
  expand_undersized_textures=false
  explicit_sync=1
  explicit_sync_kms=1
}

xwayland {
  force_zero_scaling=true
}
bind=SUPER, Return, exec, foot
bind=SUPERCONTROL, Return, exec, foot -- yazi
bind=SUPER, n, exec, neovide
bind=SUPERCONTROL, n, exec, neovide /etc/nixos
bind=SUPER, w, exec, firefox
bind=SUPER, q, killactive
bind=SUPERCONTROL, q, exec, kill -9 $(hyprctl -j activewindow | jaq -r '.pid')
bind=SUPER, Space, exec, fuzzel
bind=SUPER, x, exec, wlogout --show-binds --column-spacing 5 --row-spacing 5
bind=SUPER, d, togglespecialworkspace, terminal
bind=SUPER, m, togglespecialworkspace, mixer_tui
bind=SUPERCONTROL, m, togglespecialworkspace, mixer_gui
bind=SUPER, g, togglespecialworkspace, music_gui
bind=SUPERCONTROL, g, togglespecialworkspace, music_tui
bind=SUPER, c, togglespecialworkspace, calculator_tui
bind=SUPERCONTROL, c, togglespecialworkspace, calculator_gui
bind=SUPER, e, togglespecialworkspace, file_manager_tui
bind=SUPERCONTROL, e, togglespecialworkspace, file_manager_gui
bind=, XF86Explorer, exec, sleep 1 && hyprctl dispatch dpms off
bind=, XF86HomePage, exec, firefox
bind=, Print, exec, grimblast --notify copy output
bind=CONTROL, Print, exec, grimblast --notify --freeze copy area
bind=SHIFT, Print, exec, grimblast --freeze save area - | swayimg -
bind=ALT, Print, exec, grimblast --notify copy active
bind=SUPER, s, exec, wl-paste | satty -f -
bind=, XF86AudioPrev, exec, playerctl previous
bind=, XF86AudioNext, exec, playerctl next
bind=, XF86AudioStop, exec, playerctl position 0
bind=, XF86AudioPlay, exec, playerctl play-pause
bind=, XF86AudioMute, exec, wpctl set-mute @DEFAULT_AUDIO_SINK@ toggle
bind=SUPER, h, movefocus, l
bind=SUPER, j, movefocus, d
bind=SUPER, k, movefocus, u
bind=SUPER, l, movefocus, r
bind=SUPERSHIFT, h, movewindow, l
bind=SUPERSHIFT, j, movewindow, d
bind=SUPERSHIFT, k, movewindow, u
bind=SUPERSHIFT, l, movewindow, r
bind=SUPERSHIFT, Space, movetoworkspace, +0
bind=SUPER, 1, workspace, 1
bind=SUPER, 2, workspace, 2
bind=SUPER, 3, workspace, 3
bind=SUPER, 4, workspace, 4
bind=SUPER, 5, workspace, 5
bind=SUPER, 6, workspace, 6
bind=SUPER, 7, workspace, 7
bind=SUPER, 8, workspace, 8
bind=SUPER, 9, workspace, 9
bind=SUPER, 0, workspace, 10
bind=SUPER, Tab, workspace, previous
bind=SUPERSHIFT, 1, movetoworkspacesilent, 1
bind=SUPERSHIFT, 2, movetoworkspacesilent, 2
bind=SUPERSHIFT, 3, movetoworkspacesilent, 3
bind=SUPERSHIFT, 4, movetoworkspacesilent, 4
bind=SUPERSHIFT, 5, movetoworkspacesilent, 5
bind=SUPERSHIFT, 6, movetoworkspacesilent, 6
bind=SUPERSHIFT, 7, movetoworkspacesilent, 7
bind=SUPERSHIFT, 8, movetoworkspacesilent, 8
bind=SUPERSHIFT, 9, movetoworkspacesilent, 9
bind=SUPERSHIFT, 0, movetoworkspacesilent, 10
bind=SUPERSHIFTCONTROL, 1, movetoworkspace, 1
bind=SUPERSHIFTCONTROL, 2, movetoworkspace, 2
bind=SUPERSHIFTCONTROL, 3, movetoworkspace, 3
bind=SUPERSHIFTCONTROL, 4, movetoworkspace, 4
bind=SUPERSHIFTCONTROL, 5, movetoworkspace, 5
bind=SUPERSHIFTCONTROL, 6, movetoworkspace, 6
bind=SUPERSHIFTCONTROL, 7, movetoworkspace, 7
bind=SUPERSHIFTCONTROL, 8, movetoworkspace, 8
bind=SUPERSHIFTCONTROL, 9, movetoworkspace, 9
bind=SUPERSHIFTCONTROL, 0, movetoworkspace, 10
bind=SUPER, t, togglefloating
bind=SUPER, p, exec, /nix/store/69g2vnhskd8mgvxjxqfp4mkzkmxj9rf5-pin
bind=, F11, fullscreen, 0
bind=SUPER, f, fullscreen, 1
bind=SUPER, z, exec, /nix/store/nqs7g8hzxgma17322qw47hmrgqj49ip6-lock-cursor
bind=SUPERCONTROL, v, exec, notify-send "Playing video" "$(wl-paste)"; mpv "$(wl-paste | sed 's/&.*$//')"
bind=SUPERCONTROL, i, exec, /nix/store/gvdh96clphrah8kl0acmsl5jns6qcxai-open-image
bind=SUPERCONTROL, w, exec, notify-send "Opening link" "$(wl-paste)"; firefox "$(wl-paste)"
bind=SUPERCONTROL, x, exec, /nix/store/2r43dhf3xr483r8l7rd8nv2k6fr2qw11-zenity-4.0.3/bin/zenity --question --text 'Do you really want to reboot to Windows?' --icon system-reboot && systemctl reboot --boot-loader-entry=auto-windows
bind=SUPER, a, exec, /nix/store/r5fhaawp9cd67c1jrj68xvk05kbliqy2-notify-information
bind=SUPER, i, exec, /nix/store/frbb68xpaiw6c16nbhjk0b0jp7n8qxdr-toggle-inhibit-sleep
binde=SUPERCONTROL, h, resizeactive, -50 0
binde=SUPERCONTROL, j, resizeactive, 0 50
binde=SUPERCONTROL, k, resizeactive, 0 -50
binde=SUPERCONTROL, l, resizeactive, 50 0
binde=, XF86AudioRaiseVolume, exec, wpctl set-volume @DEFAULT_AUDIO_SINK@ 5%+
binde=, XF86AudioLowerVolume, exec, wpctl set-volume @DEFAULT_AUDIO_SINK@ 5%-
binde=, XF86MonBrightnessUp, exec, brightnessctl set 2%+
binde=, XF86MonBrightnessDown, exec, brightnessctl set 2%- -n
bindm=SUPER, mouse:272, movewindow
bindm=SUPER, mouse:273, resizewindow
env=XDG_SESSION_DESKTOP, Hyprland
env=GDK_BACKEND, wayland
env=SDL_VIDEODRIVER, wayland,x11,windows
env=QT_QPA_PLATFORM, wayland;xcb
exec-once=[workspace 1] firefox
exec-once=gammastep -l 47.1625:19.5033 -t 6500K:2600K
exec-once=/nix/store/fxx6d5m1ns1i53cjwd20gd5qnfcbqp7l-socket
exec-once=vesktop
monitor=DP-1, 1920x1080@165, 0x0, 1, vrr, 2
monitor=DP-2, 1920x1080@60, 1920x0, 1
windowrule=pin, ^(dragon)$
windowrule=float, ^(SVPManager)$
windowrule=float, ^(org.freedesktop.impl.portal.desktop.kde)$
windowrule=float, ^(xdg-desktop-portal-gtk)$
windowrule=idleinhibit always, ^(org.qbittorrent.qBittorrent)$
windowrule=noanim, ^(ueberzugpp_)(.*)$
windowrule=workspace 10, vesktop
windowrulev2=stayfocused, initialtitle:^()$, initialclass:^(steam)$
windowrulev2=minsize 1 1, initialtitle:^()$, initialclass:^(steam)$
windowrulev2=maximize, initialtitle:^(S+)$, initialclass:^(steamwebhelper)$
windowrulev2=immediate, initialclass:^(steam_app_)(.*)$
windowrulev2=fullscreen, initialclass:^(steam_app_)(.*)$
workspace=w[v1] s[false], gapsin:0, gapsout:0, border:false, shadow:false, rounding:false, decorate:false
workspace=f[1], gapsin:0, gapsout:0, border:false, shadow:false, rounding:false, decorate:false
workspace=special:terminal, on-created-empty:foot
workspace=special:mixer_gui, on-created-empty:pavucontrol
workspace=special:mixer_tui, on-created-empty:foot -- pulsemixer
workspace=special:music_gui, on-created-empty:spotify
workspace=special:music_tui, on-created-empty:foot -- spotify_player
workspace=special:calculator_gui, on-created-empty:qalculate-qt
workspace=special:calculator_tui, on-created-empty:foot -- qalc
workspace=special:file_manager_gui, on-created-empty:thunar
workspace=special:file_manager_tui, on-created-empty:foot -- yazi
workspace=1, monitor:DP-1, default:true
workspace=2, monitor:DP-1
workspace=3, monitor:DP-1
workspace=4, monitor:DP-1
workspace=5, monitor:DP-1
workspace=6, monitor:DP-1
workspace=7, monitor:DP-1
workspace=8, monitor:DP-1
workspace=9, monitor:DP-1
workspace=10, monitor:DP-2, default:true

======Config-End========


```

</details>

### Description

If a Wayland client sets its size using
```
xdg_toplevel.set_min_size(width, height + 50)
xdg_toplevel.set_max_size(width, height + 50)
```
and sets its geometry with
```
xdg_surface.set_window_geometry(0, -50, width, height + 50)
```
then that should result in a window size of `(width, height)` (which is the case in other compositors, and is expected by some things), but that's not the case, and instead, the window is size `(width, height + 50)`.

### How to reproduce

1. Run a Wayland client with the behavior explained above

I have 2 examples:

<details> <summary>Simple Raylib program</summary>

```c
#include <raylib.h>

int main(void) {
  InitWindow(200, 200, "raylibtest");
  SetTargetFPS(10);

  while (!WindowShouldClose()) {
    BeginDrawing();
    EndDrawing();
  }

  return 0;
}
```

</details>

<details> <summary>Raw Rust Wayland client (Rust because I did this a while ago, and just reused it for this example)</summary>

```rust
use std::{error::Error, io::Write, os::fd::AsFd};
use wayland_client::{
    delegate_noop,
    protocol::{
        wl_buffer::WlBuffer,
        wl_compositor::WlCompositor,
        wl_output::WlOutput,
        wl_registry::{self, WlRegistry},
        wl_shm::{Format, WlShm},
        wl_shm_pool::WlShmPool,
        wl_surface::WlSurface,
    },
    Connection, Dispatch, QueueHandle,
};
use wayland_protocols::xdg::shell::client::{
    xdg_popup::XdgPopup,
    xdg_positioner::XdgPositioner,
    xdg_surface::{self, XdgSurface},
    xdg_toplevel::{self, XdgToplevel},
    xdg_wm_base::{self, XdgWmBase},
};

struct State {
    playing: bool,

    wm_base: Option<XdgWmBase>,

    base_surface: Option<WlSurface>,
    buffer: Option<WlBuffer>,
    xdg_surface: Option<XdgSurface>,
    xdg_toplevel: Option<XdgToplevel>,
}

const WIDTH: i32 = 100;
const HEIGHT: i32 = 100;

impl State {
    fn init_xdg_surfaces(&mut self, qh: &QueueHandle<State>) {
        if self.wm_base.is_none() || self.base_surface.is_none() {
            return;
        }

        let wm_base = self.wm_base.as_ref().unwrap();
        let base_surface = self.base_surface.as_ref().unwrap();

        let xdg_surface = wm_base.get_xdg_surface(base_surface, qh, ());
        let xdg_toplevel = xdg_surface.get_toplevel(qh, ());

        xdg_toplevel.set_min_size(WIDTH, HEIGHT + 50);
        xdg_toplevel.set_max_size(WIDTH, HEIGHT + 50);
        xdg_surface.set_window_geometry(0, -50, WIDTH, HEIGHT + 50);

        xdg_toplevel.set_app_id(env!("CARGO_BIN_NAME").to_string());
        base_surface.commit();

        self.xdg_surface = Some(xdg_surface);
        self.xdg_toplevel = Some(xdg_toplevel);
    }
}

delegate_noop!(State: ignore WlCompositor);
delegate_noop!(State: ignore WlSurface);
delegate_noop!(State: ignore WlShm);
delegate_noop!(State: ignore WlShmPool);
delegate_noop!(State: ignore WlBuffer);
delegate_noop!(State: ignore WlOutput);
delegate_noop!(State: ignore XdgPositioner);
delegate_noop!(State: ignore XdgPopup);

impl Dispatch<WlRegistry, ()> for State {
    fn event(
        state: &mut Self,
        registry: &WlRegistry,
        event: wl_registry::Event,
        _: &(),
        _: &Connection,
        qh: &QueueHandle<State>,
    ) {
        if let wl_registry::Event::Global {
            name, interface, ..
        } = event
        {
            match interface.as_str() {
                "wl_compositor" => {
                    let compositor = registry.bind::<WlCompositor, _, _>(name, 6, qh, ());

                    state.base_surface = Some(compositor.create_surface(qh, ()));

                    state.init_xdg_surfaces(qh);
                }
                "wl_shm" => {
                    let shm = registry.bind::<WlShm, _, _>(name, 1, qh, ());

                    let mut file = tempfile::tempfile().unwrap();
                    let _ = file.write_all(&[0xFF, 0xFF, 0xFF, 0xFF].repeat((WIDTH * HEIGHT) as usize));
                    let stride = WIDTH * 4;
                    let pool = shm.create_pool(file.as_fd(), stride * HEIGHT, qh, ());
                    state.buffer = Some(pool.create_buffer(0, WIDTH, HEIGHT, stride, Format::Argb8888, qh, ()));

                    state.init_xdg_surfaces(qh);
                }
                "xdg_wm_base" => {
                    state.wm_base = Some(registry.bind::<XdgWmBase, _, _>(name, 5, qh, ()));

                    state.init_xdg_surfaces(qh);
                }
                _ => {}
            }
        }
    }
}

impl Dispatch<XdgWmBase, ()> for State {
    fn event(
        _: &mut Self,
        wm_base: &XdgWmBase,
        event: xdg_wm_base::Event,
        _: &(),
        _: &Connection,
        _: &QueueHandle<Self>,
    ) {
        if let xdg_wm_base::Event::Ping { serial } = event {
            wm_base.pong(serial);
        }
    }
}

impl Dispatch<XdgToplevel, ()> for State {
    fn event(
        state: &mut Self,
        _: &XdgToplevel,
        event: xdg_toplevel::Event,
        _: &(),
        _: &Connection,
        _: &QueueHandle<Self>,
    ) {
        if let xdg_toplevel::Event::Close {} = event {
            state.playing = false;
        }
    }
}

impl Dispatch<XdgSurface, ()> for State {
    fn event(
        state: &mut Self,
        xdg_surface: &XdgSurface,
        event: xdg_surface::Event,
        _: &(),
        _: &Connection,
        _: &QueueHandle<Self>,
    ) {
        if let xdg_surface::Event::Configure { serial } = event {
            xdg_surface.ack_configure(serial);

            let surface = state.base_surface.as_ref().unwrap();
            // buffer is guaranteed to be Some
            surface.attach(state.buffer.as_ref(), 0, 0);
            surface.commit();
        }
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    let mut state = State {
        playing: true,

        wm_base: None,

        base_surface: None,
        buffer: None,
        xdg_surface: None,
        xdg_toplevel: None,
    };

    let connection = Connection::connect_to_env()?;
    let display = connection.display();

    let mut queue = connection.new_event_queue::<State>();
    let qh = queue.handle();

    display.get_registry(&qh, ());

    while state.playing {
        queue.blocking_dispatch(&mut state)?;
    }

    Ok(())
}
```

</details>

### Crash reports, logs, images, videos

The Rust example above running on Hyprland vs Sway:

![image](https://github.com/user-attachments/assets/2d222011-c256-435c-9ee4-7df3f1d03eeb)

![tmp](https://github.com/user-attachments/assets/cf5f4acf-b201-4cb9-906a-38ffa20cefb4)

**Comment by vaxerski** [comment]
the C example does not produce the bug on my machine and rust example does not compile. `rustc ./main.rs` fails with a bunch of missing stuff.

Please do not expect me to know how to compile that god-awful language

**Comment by diniamo** [comment]
No idea, it happens for me with the Raylib example:

![image](https://github.com/user-attachments/assets/a4c9877b-3a46-4a6d-b17b-fa8ede90412e)

Just to clarify, the problem is that the window size isn't 200x200 (but 200x237 instead).

I'll come up with another example in C / raw Wayland if that'll help.

**Comment by diniamo** [comment]
@vaxerski alright, here is a C example: https://gist.github.com/diniamo/4d6bd22e4f1a20b76ca1ed3ea6b55d51

Note that the relevant code is way at the bottom, since I included the xdg stuff, so I could share it as one file (it should compile with `gcc main.c -l wayland-client`, as long as you have the wayland libraries and headers). I tested the example on Sway and Niri again, where the window size is still correct (200x200).

For reference, I asked another Hyprland user to run the example, and it's broken for them too.
